### PR TITLE
Drop XSS escaping because it should be aside HTML escaping. 

### LINKED ISSUE #3233: bug HTML.escape
``` ruby
irb(main):001:0> require 'cgi'
=> true
irb(main):002:0> CGI.escapeHTML("?a=1&b=2")
=> "?a=1&amp;b=2"
```

``` crystal
require "html"
p HTML.escape("?a=1&b=2")
```

`"?a&#61;1&amp;b&#61;2"`

i don't may be this is correct, i just rewrite some specs for my project from ruby and find it failed

### PR DESCRIPTION
**Pull request by akzhan**
Introduce escape_quotes mode.

Fixes #3233, refs #2175.

Thanks to @straight-shoota for clarification. Cite:

> The current implementation is not escaping HTML, but rather Javascript (inside HTML). Therefore it should be named differently (like escape_xss or escape_javascript like in Phoenix and Rails). Whether and how this should be in stdlib is a different issue.

### CODE REVIEW

File: src/html.cr [hunk 1]
```diff
@@ -28,9 +58,17 @@ module HTML
...
   #
   # HTML.escape("Crystal & You") # => "Crystal &amp; You"
+  #
+  # HTML.escape("Crystal = Me", HTML::EscapeMode::Extended) # => "Crystal &#61; Me"
   # ```
-  def self.escape(string : String) : String
-    string.gsub(SUBSTITUTIONS)
+  def self.escape(string : String, mode : EscapeMode = EscapeMode::Default) : String
+    subst = case mode
+            when EscapeMode::Default  then DEFAULT_ESCAPE
```
Comment (akzhan) [inline src/html.cr]: Default extracted as first one to optimize performance for most use case.

**Comment by akzhan** [comment]
Pull request updated (Extended renamed to XSS with documentation).

CI failed due to unrelated reason.

**Comment by RX14** [comment]
`HTML.escape` is a security-related function so I think we need a think about how this function will *actually* be used as opposed to how it's meant to be used. I think the default should be the highest escaping level because XSS is silent but escaping too much is likely to show up in development.

**Comment by akzhan** [comment]
@RX14 HTML.escape is not security-related function. https://www.w3.org/International/questions/qa-escapes#use

It is its meaning at all. All HTML-related /software depends on it.

XSS escaping is another issue.

**Comment by RX14** [comment]
Except that the reality is that people will use `HTML.escape` for XSS escaping, regardless of what the internet says.

**Comment by akzhan** [comment]
Firstly I wants to implement separate `xss_escape` method, but realizes as escaping mode option due to possibility of other flavors of escaping.

**Comment by akzhan** [comment]
Anyway previuos implementation was incorrect, because unusable in practice, see #3233, for example.

**Comment by straight-shoota** [comment]
What's the significance of escaping a slash? By itself it doesn't have any "dangerous" meaning in (X)HTML. OWASP recommends to escape it but they don't give a good argument and I can't think of any.

Single and double quotes are not necessarily harmful, either. Only if an escaped string is used inside an attribute value. But we can assume that it's better to be safe and escape them by default. I would add an option to disable escaping of quotes, though.

But: When double quotes are escaped, single quotes ought to be escaped as well. In HTML both can be used as delimiters for attributes values. I don't see any reason to escape double quotes, but not single quotes.

There are different interpretations of what is considered to be dangerous in HTML, therefore there are different sets of escape characters in common use:

* `& <>` -> PHP `htmlspecialchars` (with `ENT_NOQUOTES`), Python `cgi.escape`, W3C recommendation
* `& <> "` -> PHP `htmlspecialchars`
* `& < > " '` -> Python `html.escape`, Phoenix `Phoenix.HTML`, Go `html.EscapeString`, Django, Jinja, W3C recommendation
* `& < > " ' /` -> Ruby `CGI.escape`, PHP `htmlspecialchars` (with `ENT_QUOTES`), `Rack::Utils.escape_html`, OWASP recommendation

The default set should be as minimal as possible and as strong as necessary.
I'd suggest to use `& < > " '` as default and an option to not escape quotes (i.e. `& < >`).
XSS is an entirely different thing and should require a detailed discussion.

**Comment by akzhan** [comment]
Thanks, @straight-shoota.

Looks like we need four flavors:
* Default - `& < > " '` as Python, Go, etc. and W3C recommendation.
* CGI - `& <>` as CGI variant (it's rare case for now).
* OWASP - owasp one.
* XSS - I don't know who use it. It's current implementation.

**Comment by akzhan** [comment]
Hm, I' remove XSS escaping at all. It should be aside HTML module.

**Comment by straight-shoota** [comment]
Exactly. The current implementation is not escaping HTML, but rather Javascript (inside HTML). Therefore it should be named differently (like `escape_xss` or `escape_javascript like in Phoenix and Rails).

I'm also not sure if there should be a dedicated flavour to escape slashes. Besides Ruby/Rack there don't seem to be any big ones following the OWASP recommendation.

If there are just two sets, the API could be much simplified:
```crystal
def escape(string : String, escape_quotes = true)
  # ...
end
```

**Comment by akzhan** [comment]
I prefer enums because they extendable.

`String.compare` is example of bad decision for now, it should accept enum instead of boolean.

**Comment by akzhan** [comment]
Another option is to use Symbols instead of enum. Symbols preferable to extend stdlib functionality, because enum cannot be reopened AFAIR.

**Comment by straight-shoota** [comment]
I find the usage like `HTML.escape(string, escape_quotes: false)` much better conveys meaning than `HTML.escape(string, HTML::EscapeMode::CGI)`.
And I honestly don't know how this could be subject to further extensions anyway. I've looked at many implementations and they are all doing pretty much the same, more or less. HTML won't change significantly anytime (like ever).
Even if some sort of web framework wants do provide a different algorithm with other options, this would be a specialised case and should happen in a custom module.

**Comment by akzhan** [comment]
@straight-shoota Ok, you convinced me, Current Default will be true.

**Comment by ysbaddaden** [comment]
Issues talk about problems in `HTML.escape` when escaping URIs... but this isn't about escaping URIs (which should use `URI.encode`) but escaping HTML so an untrusted text can be inserted into a HTML structure, without breaking or hijacking the HTML layout, or worse, injecting forms or scripts.

Please detail actual cases and issues where this may problematic. Otherwise this pull request is dangerous, because it makes something that used to be secure, insecure unless manually configured.

**Comment by akzhan** [comment]
@ysbaddaden there is more disambiguation introduced by #2175. Author made escaping of JavaScript etc. inside HTML.

HTML requires escaping for minor subset of chars. And yes, every known implementation follows one of proposed cases. No XSS at all because it is not at HTML level.

**Comment by akzhan** [comment]
So this is just a clear extraction of HTML responsibility.

**Comment by ysbaddaden** [comment]
Just my point: we're doing better!

Are there actual problems to escaping XSS along with HTML? Not merely intellectual (it's not required *per se*) but practical: this is causing me issues in specific cases.

I suppose we could add an `insecure: true` parameter (so it's explicit you should never disable it) that would disable XSS escaping, but is it *really* worth it?

**Comment by akzhan** [comment]
I was pointed to this issue just with `=` char.

It was very frustrating for me.

But SO has a kind of answer - https://stackoverflow.com/a/13059657/1336858

**Comment by akzhan** [comment]
And anyway - escape is not security related thing.

**Comment by straight-shoota** [comment]
This PR does not make the application of `HTML.escape` significantly more dangerous unless it was improperly used to escape other things than HTML markup.

Of course, with this reduced set, it can not be entirely prevented that some unwanted alternations might be introduced into a string of HTML. But neither does the existing escape set.

The weak point are unquoted attributes, as explained by [OWASP](https://www.owasp.org/index.php/XSS_%28Cross_Site_Scripting%29_Prevention_Cheat_Sheet#RULE_.232_-_Attribute_Escape_Before_Inserting_Untrusted_Data_into_HTML_Common_Attributes):
> [...] developers frequently leave attributes unquoted. Properly quoted attributes can only be escaped with the corresponding quote. Unquoted attributes can be broken out of with many characters, including [space] % * + , - / ; < = > ^ and |.

To rule out every possible hijacking of unquoted attributes,, "all characters with ASCII values less than 256" except alphanumerical characters need to be escaped. Not just the few that are in the current implementation. They make it seem to be better then just those proposed in this PR, but don't add real security, because every other special character in the above code range is still available.

Implementing this strong escape strategy is highly impractical because it completely prohibits usage of those special characters inside HTML-escaped strings - think of a markup processor that prevents direct HTML input, but uses special characters in the escaped text string for HTML-based enhancements.

That's why every language library and web framework I've come across uses only the few escape characters mentioned in this PR. Unquoted attributes with user-input are highly error-prone and cannot generally be made safe without heavy impacts on functionality.

**Comment by RX14** [comment]
I still don't see why too much escaping is a bad thing, as long as it unescapes correctly. Developers will use `HTML.escape` for XSS escaping because "hey it's in a script tag in HTML I must use HTML.escape".

And yes this is absolutely a security issue when it's possible to inject arbitrary markup or even scripts into a page.

@straight-shoota What does attribute escaping have to do with this?

I agree with @ysbaddaden entirely on this. We're getting conflicting information on what escaping is needed so it's very unwise to change the secure but possibly overzealous implementation to one which might be insecure.

**Comment by straight-shoota** [comment]
@RX14 In HTML, there are only *three characters that should always appear in content as escapes* (c.f. [HTML spec](https://www.w3.org/TR/html4/charset.html#h-5.3.2), [W3C escape guidelines](https://www.w3.org/International/questions/qa-escapes#use)). **If these three are escaped, ordinary strings are perfectly safe inside regular HTML content.**

```html
<p>
  Here, you can put as many " ' / ! [ ] or whatever, you only need to escape &lt; &gt; and &amp;
</p>
````

(arguably, closing tags `>` dont't necessarily have to be escaped but it makes sense to treat them the same as opening tags `<`. and there were some issues with older browsers...)

If escaped strings are only used outside of HTML tags, this is sufficient (hence the option to only escape those three).

It gets a bit more complicated when escaped strings are to be put *inside* of HTML tags, or more precisely, attribute values. There, it is essential to escape attribute delimiters to prevent the escaped string to spill outside the attribute value limits. Valid delimiters are matching double or single quotations marks, therefor these should also be escaped.

```html
<a href="/ ! [ ] or whatever are fine, escaping is needed for &quot; &#27; &lt; &gt; and &amp;">
````

Unfortunately, the [spec](https://www.w3.org/TR/html4/intro/sgmltut.html#h-3.2.2) also allows values without delimiters and surely browsers accept them:
> In certain cases, authors may specify the value of an attribute without any quotation marks. The attribute value may only contain letters (a-z and A-Z), digits (0-9), hyphens (ASCII decimal 45), periods (ASCII decimal 46), underscores (ASCII decimal 95), and colons (ASCII decimal 58). We recommend using quotation marks even when it is possible to eliminate them.

```html
<img alt=here_you_can_only_put_alphanumericals_-_._and_:_even_a_comma_ends_this>
````

So basically, every other character would effectively be ending an unquoted attribute value and allow to inject some additional attribute. To avoid this, all these characters would have to be escaped, if the escaped string is to be used as part of an unquoted attribute value.
This would make a truely insane default, having so many escaped characters introduced when they are usually not dangerous in any way. Most escaped strings will be used as HTML content outside of tags and attribute values should preferably be always be delimited quotes.

Too much escaping is a bad thing, because it makes unnecessary changes to the string representation of the HTML code and this will break things™. For a general-purpose HTML escape function, this is unacceptable.

It would be very difficult to auto-convert URLs in HTML-escaped strings to anchor tags when slash characters are always escaped. And think about all the people that have to read HTML code filled with superfluous escape sequences all over the place.

If you want a super-secure escaping strategy for putting unescaped strings into unquoted attributes, I would not mind adding this as an option. But then it should cover all characters except those quoted in the above spec, not just an arbitrary subset. And I really don't see a valid usecase for this.

**Comment by akzhan** [comment]
Looks like @RX14  want to embed `javascript_escape`, `sanitize*`, `json_escape`, `html_escape` methods in one.

It's possible using EscapeMode (but better to separate), but anyway current implemenetation is most worst default possible (it is not HTML escape).

**Comment by straight-shoota** [comment]
Yes, these are entirely different escaping strategies for individual contexts and should not be confused with basic HTML escaping.

I've tried to dig up a reasonable explanation for escaping forward slash, but did not find an.
[Microsofts Web Protection Library](http://wpl.codeplex.com/discussions/267461) does not *consider it a dangerous character*. Someone [asked on the OWASP mailing list](https://lists.owasp.org/pipermail/esapi-dev/2012-April/002150.html) about it's significance and the only explanation was that it could be used to start a Javascript comment which could allow sneaking in other Javascript content. While this is true, it covers a different context than HTML element content. Properly escaping a string for use as a Javascript value has other requirements than just adding an HTML escape sequence for forward slash. On [StackExchange](https://security.stackexchange.com/questions/49852/owasp-escaping-recommendation-why), someone noted it is a *HTML character with special meaning* and it was detailed, that prehistoric SGML supported so-called *Null End Tag* where a forward slash is significant.

But all this has no real relevance neither for security nor sanity, and that's why apart from `Rack::Util` there does not seem to be much support for escaping forward slashes in other frameworks.

*I wrongly stated before that Ruby CGI.escape_html would escape forward slash, too. [This is not the case](https://ruby-doc.org/stdlib-1.9.3/libdoc/cgi/rdoc/CGI.html#method-c-escapeHTML).*

**Comment by straight-shoota** [comment]
`HTML.unescape` should be updated as well to perform the inversion of `HTML.escape`.

There is also `XML.escape` which should share the same implementation (unless we would decide to escape `/` for HTML as well, because this is only from HTML's past with SGML) and should be accompanied by `XML.unescape`.

**Comment by straight-shoota** [comment]
What's holding this up? I am confident the reasons for this PR have been properly explained.

**Comment by ysbaddaden** [comment]
Let's sum up:
- pro-reasons: standards say you're being over zealous.
- con-reasons: better safe than sorry.

Let's close.

**Comment by ysbaddaden** [comment]
Let's be constructive:

1. This pull request drops a *security* feature without providing *any* replacement. We MUST introduce a solution to escape other XSS possibilities, namely `HTML.escape_javascript` for example, following [OWASP rule #3](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#RULE_.233_-_JavaScript_Escape_Before_Inserting_Untrusted_Data_into_JavaScript_Data_Values). 

2. This pull request introduces different escaping strategies, for *no* reason. We MUST follow [OWASP rule #2](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#RULE_.231_-_HTML_Escape_Before_Inserting_Untrusted_Data_into_HTML_Element_Content), so `HTML.escape` MUST always escape the following characters:

```
&  --> &amp;
<  --> &lt;
>  --> &gt;
"  --> &quot;
'  --> &#x27;
/  --> &#x2F;
```

3. let's go the extra mile, and provide a `HTML.escape_attribute` following [OWASP rule #2](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#RULE_.232_-_Attribute_Escape_Before_Inserting_Untrusted_Data_into_HTML_Common_Attributes) recommendation.

4. Last but not least: we MUST document the breaking change as a *security* issue in the CHANGELOG.

**Comment by akzhan** [comment]
ok, somebody should add `escape_javascript` and `escape_attribute`. And drop `escape_quotes` mode.

**Comment by straight-shoota** [comment]
1. I disagree: The current implementation is just *buggy* and as far as I can tell this is not because it is intended to a *security* feature. There is no direct replacement and therefore we can't provide one (because it's just an incorrect implementation). True, we SHOULD add additional escape methods for other contexts but I don't think this MUST necessarily be included in this PR.

2. Maybe `escape_javascript` should rather be more like [Rails's version](http://api.rubyonrails.org/classes/ActionView/Helpers/JavaScriptHelper.html#method-i-escape_javascript) than the OWASP recommendation? I'm not sure. But I'd recommend to address this in another PR.

3. I really don't see any need for this: Having a special escape method for attribute values is only necessary if attributes are written without quotes. Otherwise `HTML.escape` is perfectly fine. I don't think we should encourage the use of unquoted attribute.

4. There is a reason for this: If the content returned from `HTML.escape` is to be used inside a HTML tag like `<a <%= HTML.escape(attrs, false) %>>` (because you still need to escape `< > &`) it would be *wrong* to escape quotes, because `href="#"` would become `<a href=&quot;#&quot;>` instead of `<a href="#">`. But the default should be to escape quotes.

5. As far as I can tell, thus far the changelog has only been updated when there was a new release. I'm not aware if this practice has changed. If not, this should probably be discussed elsewhere.

**Comment by ysbaddaden** [comment]
Endless arguing never leads anywhere. Please implement *at least* the following and I'll **happily** merge the pull request:

* `HTML.escape(unsafe)` always escapes *all* `& < > " ' /` characters —injecting raw HTML as attributes without escaping quotes doesn't prevent from injected quotes inside attribute values, it's thus insecure, and shouldn't be unsupported.
* `HTML.escape_javascript(unsafe)` with a secure enough list of escaped characters.

**Comment by straight-shoota** [comment]
`HTML.escape(unsafe)` is already there. Do you really think it hurts to add an option for not escaping quotes? This is after all the [recommendation from W3C](https://www.w3.org/International/questions/qa-escapes#use).

For `HTML.escape_javascript(unsafe)` there are a number of possible implementations, just a few examples:
* [OWASP Rule #3](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#RULE_.233_-_JavaScript_Escape_Before_Inserting_Untrusted_Data_into_JavaScript_Data_Values): *Except for alphanumeric characters, escape all characters less than 256 with the \xHH format*
* [Phoenix.HTML.escape_javascript](https://hexdocs.pm/phoenix_html/Phoenix.HTML.html#escape_javascript/1):  *Escapes quotes (double and single), double backslashes and other.* `</ \ ' " \r \n \r\n \u2028 \u2029`
* [Rails ActionView::Helpers::JavaScriptHelper.escape_javascript](http://api.rubyonrails.org/classes/ActionView/Helpers/JavaScriptHelper.html#method-i-escape_javascript): *Escapes carriage returns and single and double quotes for JavaScript segments.* `</  " ' \n \r  \r\n \342\200\250 \342\200\251`
* [Django escapejs](https://docs.djangoproject.com/en/1.11/ref/templates/builtins/#escapejs): *Escapes characters for use in JavaScript strings.* `\ ' " > < & = - ; \u2028 \u2029`
* [Symfony tag helper escape_javascript](http://www.symfony-project.org/api/1_0/TagHelper#method_escape_javascript): *Escape carrier returns and single and double quotes for Javascript segments.*: `" ' \r \n \r\n`

I honestly don't know which one should be implemented. Though I'd tend to the direction of Rails and Phoenix. They prohibit to break out of HTML tag content or attribute contexts as well as Javascript strings. That should be all that is important for this.

**Comment by akzhan** [comment]
@ysbaddaden I have added `escape_javascript`, but should mention that it's escaping sequences are very different than HTML ones (any known implementation).

File: src/html.cr [hunk 2]
```diff
@@ -38,13 +49,48 @@ module HTML
...
       io << subst.fetch(char, char)
     end
   end
 
+  # Encodes a string with JavaScript escaping substitutions.
+  #
+  # ```
+  # require "html"
+  #
+  # HTML.escape_javascript("</crystal> \u2028") # => ""<\\/crystal> &#x2028;""
```
Comment (straight-shoota) [inline src/html.cr]: double quotes should be reduced.

Comment (akzhan) [reply src/html.cr]: fixed, thanks

File: src/html.cr [hunk 3]
```diff
@@ -40,9 +49,45 @@ module HTML
...
+  # HTML.escape_javascript("</crystal> \u2028") # => "<\\/crystal> &#x2028;"
+  # ```
+  def self.escape_javascript(string : String) : String
+    string.gsub("\r\n", "\n").gsub(ESCAPE_JAVASCRIPT_SUBST).gsub("</", "<\\/")
+  end
+
+  # Encodes a string with JavaScript escaping, but writes to the `IO` instance provided.
+  #
+  # ```
+  # io = IO::Memory.new
```
Comment (straight-shoota) [inline src/html.cr]: It'd be better to use a string builder as example.

File: src/html.cr [hunk 4]
```diff
@@ -40,9 +49,45 @@ module HTML
...
+  # HTML.escape_javascript("</crystal> \u2028", io) # => nil
+  # io.to_s                                         # => "<\\/crystal> &#x2028;"
+  # ```
+  def self.escape_javascript(string : String, io : IO) : Nil
+    previous_char = '\0'
     string.each_char do |char|
-      io << SUBSTITUTIONS.fetch(char, char)
+      if previous_char == '\r' && char == '\n'
```
Comment (straight-shoota) [inline src/html.cr]: Maybe this would be better as a `case` statement to express control flow more clearly and be shorter overall.
```crystal
case
when previous_char == '\r' && char == '\n'
when previous_char == '<' && char == '/'
  io << '\\' << '/'
else
  io << ESCAPE_JAVASCRIPT_SUBST.fetch(char, char)
end
previous_char = char
```

File: spec/std/html_spec.cr [hunk 5]
```diff
@@ -10,21 +10,41 @@ describe "HTML" do
     end
 
     it "escapes dangerous characters from a string" do
```
Comment (straight-shoota) [inline spec/std/html_spec.cr]: Maybe reword to `escapes special characters from HTML string`

File: spec/std/html_spec.cr [hunk 6]
```diff
@@ -10,21 +10,41 @@ describe "HTML" do
...
+  end
+
+  describe ".escape_javascript" do
+    it "does not change a safe string" do
+      str = HTML.escape_javascript("safe_string")
+
+      str.should eq("safe_string")
+    end
+
+    it "escapes dangerous characters from a string" do
```
Comment (straight-shoota) [inline spec/std/html_spec.cr]: Maybe reword `escape special characters from a JavaScript string`? The characters are not "dangerous" per se, they just might convey special meaning.

File: src/html.cr [hunk 7]
```diff
@@ -29,8 +38,8 @@ module HTML
   #
   # HTML.escape("Crystal & You") # => "Crystal &amp; You"
   # ```
```
Comment (straight-shoota) [inline src/html.cr]: It would be nice to list all character substitutions explicitly, so you don't need to look at the source code to figure out what exactly gets escaped and why.

Suggestion:

```crystal
# Escapes `&`, `<`,  `>`, `"` and `'` chars as `&amp;`, `&lt;`, `&gt;`, `&quot;` and `&#27` according to 
# [OWASP Rule #1](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#RULE_.231_-_HTML_Escape_Before_Inserting_Untrusted_Data_into_HTML_Element_Content).
# If *escape_quotes* is `false`, `"` and `'` will not be escaped ([W3C recommendation](https://www.w3.org/International/questions/qa-escapes#use)).
```

The other methods should have similar explanations.

File: spec/std/html_spec.cr (hunk 5 above)
Comment (akzhan) [reply spec/std/html_spec.cr]: Done, thanks.

File: spec/std/html_spec.cr (hunk 6 above)
Comment (akzhan) [reply spec/std/html_spec.cr]: Done, thanks.

**Comment by straight-shoota** [comment]
@akzhan What do you think about `XML.escape`? As [described above](https://github.com/crystal-lang/crystal/pull/4555#issuecomment-311755439) I think it makes no sense to have both `XML.escape` and `HTML.escape` wich do the same thing.

Since XML is more general, should we put the method only there? But most people are actually using HTML, so maybe it would be better to have it (only) there. Or make an exception and allow aliases in this case.

It wouldn't be the same if HTML version would also escape `/` but that doesn't make any sense despite being an OWASP recommendation.

**Comment by straight-shoota** [comment]
And perhaps we should think about making escaping `'` optional as well. It's probably not advisable, but certain specifications expect that single quotes are not escaped, for example [Common Mark *Textual content*](http://spec.commonmark.org/0.27/#example-620). Libxml's `xmlEncodeSpecialChars` also does not escape single quotes (see [source](https://github.com/tenderlove/libxml2/blob/master/entities.c#L708)). But for some reason it escapes `\r` as `&#13;`.

I was looking at markd's source code where there is a custom [`escape` method](https://github.com/icyleaf/markd/blob/57ba89d698661da69368b6b98fef0799643ad62e/src/markd/renderer.cr#L29). I figured it could be replaced with the implementation from this PR but if single quotes are escaped, the Common Mark specs won't match.

**Comment by akzhan** [comment]
I'm tired with this PR. Feel free to propose your own.

**Comment by asterite** [comment]
@akzhan Sorry you got tired.

But remember: you don't have to do what everyone says here. You can accept suggestions from anyone, but you only have to make changes that are requested by core team members (and even then you can argue about your decision, in many cases core team members changed their mind).

I sent https://github.com/crystal-lang/crystal/pull/5012 to fix this situation.

File: src/html.cr [hunk 8]
```diff
@@ -19,6 +19,15 @@ module HTML
...
   }
+  ESCAPE_JAVASCRIPT_SUBST = {
+    '\''     => "\\'",
+    '"'      => "\\\"",
+    '\\'     => "\\\\",
+    '\u2028' => "&#x2028;",
+    '\u2029' => "&#x2029;",
+    '\n'     => "\\n",
+    '\r'     => "\\n",
+  }
```
Comment (rdp) [inline src/html.cr]: django appears to escape slightly more? https://docs.djangoproject.com/en/1.10/_modules/django/utils/html/#escape Just asking... :)

Comment (straight-shoota) [reply src/html.cr]: This PR has been closed and #5012 was merged instead. Currently there is no escape method for javascript.
